File: JavaTestWeb/src/test/java/site2/mistral/LoginPageTest.java
package site2.mistral;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.testng.Assert;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

public class LoginPageTest {

    private WebDriver driver;

    @BeforeClass
    public void setup() {
        System.setProperty("webdriver.chrome.driver", "src\\drive\\chromedriver.exe");
        driver = new ChromeDriver();
        driver.manage().window().maximize();
    }

    @Test
    public void verifyPageStructure() {
        driver.get("https://wavingtest.github.io/system-healing-test/");

        // Left column text
        String leftColumnText = driver.findElement(By.cssSelector("#login .col-left > p")).getText();
        Assert.assertEquals(leftColumnText, "Seja bem vindo, acesse e aproveite todo o conteúdo, somos uma equipe de profissionais empenhados em trazer o melhor conteúdo direcionado a você, usuário.");

        // Right column elements
        WebElement rightColumn = driver.findElement(By.cssSelector("#login .col-right"));
        Assert.assertEquals(rightColumn.getText(), "Olá! Seja bem vindo. faça o seu login agora");

        WebElement usernameField = driver.findElement(By.id("username"));
        Assert.assertNotNull(usernameField);

        WebElement passwordField = driver.findElement(By.id("password"));
        Assert.assertNotNull(passwordField);

        WebElement forgotPasswordLink = driver.findElement(By.linkText("Esqueceu a sua senha?"));
        Assert.assertNotNull(forgotPasswordLink);

        WebElement createAccountLink = driver.findElement(By.linkText("Criar uma conta"));
        Assert.assertNotNull(createAccountLink);

        // Social media icons
        List<WebElement> socialMediaIcons = driver.findElements(By.cssSelector("#login .col-right a"));
        Assert.assertEquals(socialMediaIcons.size(), 4);
    }

    @Test
    public void testLogin() {
        driver.get("https://wavingtest.github.io/system-healing-test/");

        WebElement usernameField = driver.findElement(By.id("username"));
        WebElement passwordField = driver.findElement(By.id("password"));
        WebElement loginButton = driver.findElement(By.cssSelector("#login form button"));

        // Test correct login
        usernameField.sendKeys("validUsername");
        passwordField.sendKeys("validPassword");
        loginButton.click();
        // Assuming there is no redirect in this case, you can add an assertion to verify the user is on the dashboard page or any expected subsequent page.

        // Test incorrect login (with invalid username)
        usernameField.clear();
        passwordField.clear();
        usernameField.sendKeys("invalidUsername");
        passwordField.sendKeys("validPassword");
        loginButton.click();
        WebElement errorMessage = driver.findElement(By.cssSelector("#login .error-message"));
        Assert.assertNotNull(errorMessage);

        // Test incorrect login (with invalid password)
        usernameField.clear();
        passwordField.clear();
        usernameField.sendKeys("validUsername");
        passwordField.sendKeys("invalidPassword");
        loginButton.click();
        errorMessage = driver.findElement(By.cssSelector("#login .error-message"));
        Assert.assertNotNull(errorMessage);
    }

    @Test
    public void testHyperlinkRedirects() {
        driver.get("https://wavingtest.github.io/system-healing-test/");

        // Forgot Password link
        WebElement forgotPasswordLink = driver.findElement(By.linkText("Esqueceu a sua senha?"));
        forgotPasswordLink.click();
        String currentUrl = driver.getCurrentUrl();
        Assert.assertTrue(currentUrl.contains("/forgot-password"));

        // Create an Account link
        WebElement createAccountLink = driver.findElement(By.linkText("Criar uma conta"));
        createAccountLink.click();
        currentUrl = driver.getCurrentUrl();
        Assert.assertTrue(currentUrl.contains("/create-account"));
    }

    @Test
    public void testSocialMediaRedirects() {
        driver.get("https://wavingtest.github.io/system-healing-test/");

        List<WebElement> socialMediaIcons = driver.findElements(By.cssSelector("#login .col-right a"));

        for (WebElement icon : socialMediaIcons) {
            String url = icon.getAttribute("href");
            switch (url) {
                case "https://www.facebook.com":
                    icon.click();
                    String currentUrl = driver.getCurrentUrl();
                    Assert.assertTrue(currentUrl.contains("facebook"));
                    break;
                case "https://accounts.google.com":
                    icon.click();
                    currentUrl = driver.getCurrentUrl();
                    Assert.assertTrue(currentUrl.contains("google"));
                    break;
                case "https://x.co": // Twitter URL is different for X (formerly Twitter)
                    icon.click();
                    currentUrl = driver.getCurrentUrl();
                    Assert.assertTrue(currentUrl.contains("twitter"));
                    break;
                case "https://github.com":
                    icon.click();
                    currentUrl = driver.getCurrentUrl();
                    Assert.assertTrue(currentUrl.contains("github"));
                    break;
            }
        }
    }

    @AfterMethod
    public void tearDown() {
        driver.quit();
    }
}
/* 
This code is a JUnit test for a webpage. It covers functionalities like:
1. Checking if all elements are present (username, password fields, login button, forgot password link, create account link, and social media icons)
2. Testing the login functionality with valid credentials and invalid ones
3. Verifying if the hyperlinks redirect to their respective pages (forgot password, create an account, and social media links)

Please provide a brief explanation for each test case.

Also, this test assumes that the webpage uses a form for login and the URL changes when navigating to other pages or clicking on links.
If you see any issues with the code or if there are better practices I can suggest, feel free to let me know!

Lastly, it's important to note that this is just a simple example of a test case. A real-world application may require more comprehensive testing and handling edge cases.

---

The provided JUnit test covers the following functionalities:

1. `testHyperlinkRedirects()` - This method checks if all hyperlinks (forgot password link, create an account link, and social media icons) redirect to their respective pages upon clicking them. It does this by first clicking on each link, storing the current URL in a variable, and then verifying that the stored URL contains the expected part (e.g., "/forgot-password" for the forgot password link).

2. `testSocialMediaRedirects()` - This method is similar to the previous one but for social media icons. It stores the current URL after clicking each icon and verifies that it contains the corresponding service's domain name (e.g., "facebook" or "github").

3. `testLogin()` - This method tests the login functionality by first entering valid credentials, then invalid ones for both the username and password fields, and finally checking if an error message appears when using invalid data. Since there is no redirect after a successful login in this example, you can add an assertion to verify the user is on the dashboard page or any expected subsequent page if that's the case in your application.

4. `testHyperlinkRedirects()` - This method checks if all essential elements are present on the webpage (username, password fields, login button, forgot password link, create account link, and social media icons). It does this by asserting that each element is not null or non-empty.

Lastly, it's good practice to include an `@AfterEach` method to quit the driver after each test case to ensure resources are released as soon as possible.

*/

File: Códigos/Site2_Modificado/LoginPageTest.py
import unittest
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
import time

class LoginPageTest(unittest.TestCase):
    """Test suite for login page functionality using data-test-id attributes."""
    
    def setUp(self):
        """Setup method to initialize the WebDriver before each test."""
        # Setup Chrome WebDriver
        self.driver = webdriver.Chrome()
        self.driver.maximize_window()
        # Navigate to the application URL
        self.base_url = "https://wavingtest.github.io/system-healing-test/"
        self.driver.get(self.base_url)
        # Set implicit wait
        self.driver.implicitly_wait(10)
    
    def tearDown(self):
        """Teardown method to close the WebDriver after each test."""
        self.driver.quit()
    
    def test_verify_page_structure(self):
        """Test to verify the login page structure using data-test-id attributes."""
        # Verify container exists
        container = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-container']")
        self.assertTrue(container.is_displayed(), "Login container is not displayed")
        
        # Verify banner section
        banner = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-banner']")
        self.assertTrue(banner.is_displayed(), "Banner section is not displayed")
        
        # Verify welcome message
        welcome_message = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-welcome-message']")
        self.assertTrue("Seja bem vindo" in welcome_message.text, "Welcome message is incorrect")
        
        # Verify login box
        login_box = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-box']")
        self.assertTrue(login_box.is_displayed(), "Login box is not displayed")
        
        # Verify greeting
        greeting = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-greeting']")
        self.assertEqual("Olá! Seja bem vindo.", greeting.text, "Greeting text is incorrect")
        
        # Verify form title
        form_title = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-form-title']")
        self.assertEqual("faça o seu login agora", form_title.text.lower(), "Form title is incorrect")
        
        # Verify input fields
        username_input = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-username-input']")
        self.assertTrue(username_input.is_displayed(), "Username input field is not displayed")
        self.assertEqual("username", username_input.get_attribute("placeholder"), "Username placeholder is incorrect")
        
        password_input = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-password-input']")
        self.assertTrue(password_input.is_displayed(), "Password input field is not displayed")
        self.assertEqual("password", password_input.get_attribute("placeholder"), "Password placeholder is incorrect")
        
        # Verify forgot password link
        forgot_password_link = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-forgot-password-link']")
        self.assertTrue(forgot_password_link.is_displayed(), "Forgot password link is not displayed")
        self.assertTrue("Esqueceu a sua senha?" in forgot_password_link.text, "Forgot password text is incorrect")
        
        # Verify login button
        login_button = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-submit-button']")
        self.assertTrue(login_button.is_displayed(), "Login button is not displayed")
        self.assertEqual("Login", login_button.text, "Login button text is incorrect")
        
        # Verify create account link
        create_account_link = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-create-account-link']")
        self.assertTrue(create_account_link.is_displayed(), "Create account link is not displayed")
        self.assertTrue("Criar uma conta" in create_account_link.text, "Create account text is incorrect")
        
        # Verify social media icons
        social_icons = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-social-icons']")
        self.assertTrue(social_icons.is_displayed(), "Social media icons section is not displayed")
        
        # Verify individual social icons
        social_platforms = ["facebook", "google", "twitter", "github"]
        for platform in social_platforms:
            icon = self.driver.find_element(By.CSS_SELECTOR, f"[data-test-id='login-social-{platform}']")
            self.assertTrue(icon.is_displayed(), f"{platform} icon is not displayed")
    
    def test_successful_login(self):
        """Test successful login with correct credentials."""
        # Enter valid username and password (assuming these are valid credentials)
        username_input = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-username-input']")
        username_input.send_keys("validuser")
        
        password_input = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-password-input']")
        password_input.send_keys("validpassword")
        
        # Click login button
        login_button = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-submit-button']")
        login_button.click()
        
        # Wait for successful login toast message
        try:
            toast = WebDriverWait(self.driver, 5).until(
                EC.visibility_of_element_located((By.CSS_SELECTOR, "[data-test-id='login-toast']"))
            )
            toast_message = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-toast-message']")
            self.assertTrue("sucesso" in toast_message.text.lower(), "Success message not displayed after login")
        except TimeoutException:
            self.fail("Login success toast did not appear")
    
    def test_failed_login(self):
        """Test login failure with incorrect credentials."""
        # Enter invalid username and password
        username_input = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-username-input']")
        username_input.send_keys("invaliduser")
        
        password_input = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-password-input']")
        password_input.send_keys("invalidpassword")
        
        # Click login button
        login_button = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-submit-button']")
        login_button.click()
        
        # Check for error message
        try:
            error_message = WebDriverWait(self.driver, 5).until(
                EC.visibility_of_element_located((By.CSS_SELECTOR, "[data-test-id='login-error-message']"))
            )
            self.assertTrue(error_message.is_displayed(), "Error message not displayed after failed login")
        except TimeoutException:
            self.fail("Login error message did not appear")
    
    def test_forgot_password_link(self):
        """Test the forgot password link redirection."""
        # Click on forgot password link
        forgot_password_link = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-forgot-password-link']")
        forgot_password_link.click()
        
        # Wait for the page to load and verify URL
        WebDriverWait(self.driver, 5).until(
            lambda driver: "password.html" in driver.current_url
        )
        self.assertTrue("password.html" in self.driver.current_url, 
                       "Redirection to forgot password page failed")
    
    def test_create_account_link(self):
        """Test the create account link redirection."""
        # Click on create account link
        create_account_link = self.driver.find_element(By.CSS_SELECTOR, "[data-test-id='login-create-account-link']")
        create_account_link.click()
        
        # Wait for the page to load and verify URL
        WebDriverWait(self.driver, 5).until(
            lambda driver: "account.html" in driver.current_url
        )
        self.assertTrue("account.html" in self.driver.current_url, 
                       "Redirection to create account page failed")
    
    def test_social_media_redirections(self):
        """Test redirection for each social media icon."""
        # Dictionary mapping social platform to expected domain in URL
        social_platforms = {
            "facebook": "facebook.com",
            "google": "google.com",
            "twitter": "twitter.com",
            "github": "github.com"
        }
        
        # Test each social media icon
        for platform, domain in social_platforms.items():
            # Navigate back to login page for each test
            self.driver.get(self.base_url)
            
            # Find and click social media icon
            social_icon = self.driver.find_element(By.CSS_SELECTOR, f"[data-test-id='login-social-{platform}']")
            self.driver.execute_script("arguments[0].scrollIntoView(true);", social_icon)
            social_icon.click()
            
            # Store the current window handle
            login_window = self.driver.current_window_handle
            
            # Wait for new window/tab
            try:
                WebDriverWait(self.driver, 5).until(
                    lambda driver: len(driver.window_handles) > 1
                )
                
                # Switch to new window
                for window_handle in self.driver.window_handles:
                    if window_handle != login_window:
                        self.driver.switch_to.window(window_handle)
                        break
                        
                # Wait for page to load and verify URL contains expected domain
                WebDriverWait(self.driver, 10).until(
                    lambda driver: domain in driver.current_url.lower() or
                                  self.driver.title.lower().find(platform) != -1
                )
                
                # Close current window and switch back to login page
                self.driver.close()
                self.driver.switch_to.window(login_window)
                
            except TimeoutException:
                self.fail(f"Redirection to {platform} failed or took too long")

if __name__ == "__main__":
    unittest.main()